ipykernel/iostream.py

```diff
@@ -63,24 +59,33 @@ def __init__(self, socket, pipe=False):
         self.background_socket = BackgroundSocket(self)
         self._master_pid = os.getpid()
         self._pipe_flag = pipe
-        self.io_loop = IOLoop(make_current=False)
+        self.io_loop = None
+        self.loop_ready = threading.Event()
+        self.thread = threading.Thread(target=self._thread_main, name="IOPub")
+        self.thread.daemon = True
+        self.thread.pydev_do_not_trace = True
+        self.thread.is_pydev_daemon_thread = True
+        self.thread.name = "IOPub"
+        self.thread.start()
         if pipe:
             self._setup_pipe_in()
         self._local = threading.local()
         self._events = deque()
         self._event_pipes = WeakSet()
         self._setup_event_pipe()
-        self.thread = threading.Thread(target=self._thread_main, name="IOPub")
-        self.thread.daemon = True
-        self.thread.pydev_do_not_trace = True
-        self.thread.is_pydev_daemon_thread = True
-        self.thread.name = "IOPub"
+        # make sure we don't prevent process exit
+        # I'm not sure why setting daemon=True above isn't enough, but it doesn't appear to be.
+        atexit.register(self.stop)
 
     def _thread_main(self):
         """The inner loop that's actually run in a thread"""
-        self.io_loop.make_current()
-        self.io_loop.start()
-        self.io_loop.close(all_fds=True)
+        self.io_loop = asyncio.new_event_loop()
```

Comment:
Can't you create the loop in the init (as done before) and only setting it here? This will avoid using the Event, isn't it?

Reply:
I thought that the event loop had to be created inside the thread, but it seems to work when it is created outside. Indeed the Event is not necessary then. Thanks, I'll make the change.